refactor(language-service): enable host binding features

crisbeto opened the pull request:
Reworks the language service to enable features like quick info and definitions inside host bindings.

**Note:** for the full range of language service features to work, the changes from https://github.com/angular/vscode-ng-language-service/pull/2155 have to be released.

devversion commented on packages/language-service/src/template_target.ts:
```diff
@@ -673,3 +677,22 @@ function getSpanIncludingEndTag(ast: TmplAstNode) {
 }
+
+/**
+ * Checks whether a position is within the host bindings of a directive.
```
Nit: seems like this also works for non-host binding elements

devversion commented on packages/language-service/src/utils/index.ts:
```diff
@@ -136,21 +136,43 @@ function getInlineTypeCheckInfoAtPosition(
+
+    for (const binding of resources.hostBindings) {
+      if (
+        !isExternalResource(binding) &&
```
Can you quickly refresh me on when a host binding can be an "external resource"?

crisbeto replied:
Right now it can't, because we don't produce the AST for bindings defined outside of the `host` object (e.g. through a spread operation). In theory it could be external if we started supporting type checking on dynamic `host` objects.

devversion commented on packages/language-service/test/completions_spec.ts:
```diff
@@ -2057,6 +2057,63 @@ describe('completions', () => {
+      expectContain(completions, ts.ScriptElementKind.memberVariableElement, ['title', 'hero']);
+    });
+
+    it('should be able to complete in side `host` of a directive', () => {
```
```suggestion
    it('should be able to complete inside `host` of a directive', () => {
```

devversion commented on packages/language-service/test/references_and_rename_spec.ts:
```diff
@@ -1497,6 +1497,127 @@ describe('find references and rename locations', () => {
+      expect(renameLocations.length).toBe(2);
+
+      assertFileNames(renameLocations, ['app.ts']);
+      assertTextSpans(renameLocations, ['getCompId']);
```
Is there a way for us to better assert that one is properly mapped into the host bindings?

crisbeto replied:
Hmm, I guess we could check the actual start/end of the span, but I feel like that'll make the test a bit brittle because the position can moved depending on the formatting.

devversion reviewed (approved):
LGTM, very clean! — just some nits/questions

crisbeto committed:
refactor(language-service): enable host binding features

Reworks the language service to enable features like quick info and definitions inside host bindings.

**Note:** for the full range of language service features to work, the changes from https://github.com/angular/vscode-ng-language-service/pull/2155 have to be released.

pkozlowski-opensource commented:
This PR was merged into the repository by commit 1dba2a89b070537ab3d1c8ef14e467f6fda3b969.

The changes were merged into the following branches: main
